File: shellheim/src/types/keymap.ts
// Keymap types for keyboard shortcuts

export interface Keymap {
  id: string;
  account_id: string;
  action: string;
  key: string;
  modifiers: string;
  description: string;
  enabled: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreateKeymapRequest {
  action: string;
  key: string;
  modifiers: string;
  description?: string;
}

export interface UpdateKeymapRequest {
  key?: string;
  modifiers?: string;
  enabled?: boolean;
}

export interface DefaultKeymap {
  action: string;
  key: string;
  modifiers: string;
  description: string;
}

// Available actions for keymaps
export const KEYMAP_ACTIONS = [
  "command_palette",
  "new_connection",
  "close_tab",
  "next_tab",
  "prev_tab",
  "toggle_sidebar",
  "search_servers",
  "open_sftp",
  "open_snippets",
  "disconnect",
  "copy_terminal",
  "paste_terminal",
] as const;

export type KeymapAction = (typeof KEYMAP_ACTIONS)[number];

// Human-readable action labels
export const ACTION_LABELS: Record<KeymapAction, string> = {
  command_palette: "Command Palette",
  new_connection: "New Connection",
  close_tab: "Close Tab",
  next_tab: "Next Tab",
  prev_tab: "Previous Tab",
  toggle_sidebar: "Toggle Sidebar",
  search_servers: "Search Servers",
  open_sftp: "Open SFTP",
  open_snippets: "Open Snippets",
  disconnect: "Disconnect Session",
  copy_terminal: "Copy Selection",
  paste_terminal: "Paste",
};

// Modifier key options
export const MODIFIER_OPTIONS = [
  { value: "", label: "None" },
  { value: "ctrl", label: "Ctrl" },
  { value: "shift", label: "Shift" },
  { value: "alt", label: "Alt" },
  { value: "meta", label: "⌘ / Win" },
  { value: "ctrl+shift", label: "Ctrl + Shift" },
  { value: "ctrl+alt", label: "Ctrl + Alt" },
  { value: "ctrl+shift+alt", label: "Ctrl + Shift + Alt" },
  { value: "meta+shift", label: "⌘ + Shift" },
];

// Common key options
export const KEY_OPTIONS = [
  // Letters
  ..."abcdefghijklmnopqrstuvwxyz".split(""),
  // Numbers
  ..."0123456789".split(""),
  // Function keys
  "F1",
  "F2",
  "F3",
  "F4",
  "F5",
  "F6",
  "F7",
  "F8",
  "F9",
  "F10",
  "F11",
  "F12",
  // Special keys
  "Tab",
  "Enter",
  "Space",
  "Backspace",
  "Delete",
  "Escape",
  "ArrowUp",
  "ArrowDown",
  "ArrowLeft",
  "ArrowRight",
  "Home",
  "End",
  "PageUp",
  "PageDown",
  "`",
  "-",
  "=",
  "[",
  "]",
  "\\",
  ";",
  "'",
  ",",
  ".",
  "/",
];

/**
 * Format a keymap for display (e.g., "Ctrl + P")
 */
export function formatKeyBinding(keymap: Keymap | DefaultKeymap): string {
  const parts: string[] = [];

  if (keymap.modifiers) {
    const mods = keymap.modifiers.split("+").map((m) => {
      switch (m.toLowerCase()) {
        case "ctrl":
          return "Ctrl";
        case "shift":
          return "Shift";
        case "alt":
          return "Alt";
        case "meta":
          return navigator.platform.includes("Mac") ? "⌘" : "Win";
        default:
          return m;
      }
    });
    parts.push(...mods);
  }

  // Format key
  let keyDisplay = keymap.key;
  if (keymap.key === " " || keymap.key === "Space") {
    keyDisplay = "Space";
  } else if (keymap.key.length === 1) {
    keyDisplay = keymap.key.toUpperCase();
  }
  parts.push(keyDisplay);

  return parts.join(" + ");
}

/**
 * Check if a keyboard event matches a keymap
 */
export function matchesKeymap(
  event: KeyboardEvent,
  keymap: Keymap | DefaultKeymap
): boolean {
  // DefaultKeymap doesn't have enabled property, treat as enabled
  const isEnabled = 'enabled' in keymap ? keymap.enabled : true;
  if (!isEnabled) return false;

  // Check key
  const eventKey = event.key.toLowerCase();
  const keymapKey = keymap.key.toLowerCase();
  if (eventKey !== keymapKey) return false;

  // Check modifiers
  const mods = keymap.modifiers.toLowerCase().split("+").filter(Boolean);
  const hasCtrl = mods.includes("ctrl");
  const hasShift = mods.includes("shift");
  const hasAlt = mods.includes("alt");
  const hasMeta = mods.includes("meta");

  // On Mac, treat ctrl as meta for common shortcuts
  const isMac = navigator.platform.includes("Mac");
  const ctrlOrMeta = isMac ? event.metaKey : event.ctrlKey;

  if (hasCtrl && !ctrlOrMeta) return false;
  if (hasShift && !event.shiftKey) return false;
  if (hasAlt && !event.altKey) return false;
  if (hasMeta && !event.metaKey) return false;

  // Ensure no extra modifiers are pressed
  if (!hasCtrl && !hasMeta && (event.ctrlKey || (isMac && event.metaKey)))
    return false;
  if (!hasShift && event.shiftKey && keymap.key.length === 1) return false;
  if (!hasAlt && event.altKey) return false;

  return true;
}

File: shellheim/src-tauri/src/models/keymap.rs
//! Keymap model for customizable keyboard shortcuts

use serde::{Deserialize, Serialize};

/// Keymap entry - stores user-defined keyboard shortcuts
#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct Keymap {
    pub id: String,
    pub account_id: String,
    pub action: String,        // e.g., "command_palette", "new_connection", "close_tab"
    pub key: String,           // e.g., "p", "n", "w"
    pub modifiers: String,     // e.g., "ctrl", "ctrl+shift", "meta"
    pub description: String,   // Human-readable description
    pub enabled: bool,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateKeymapRequest {
    pub action: String,
    pub key: String,
    pub modifiers: String,
    pub description: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateKeymapRequest {
    pub key: Option<String>,
    pub modifiers: Option<String>,
    pub enabled: Option<bool>,
}

/// Default keymaps that ship with the application
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DefaultKeymap {
    pub action: String,
    pub key: String,
    pub modifiers: String,
    pub description: String,
}

impl DefaultKeymap {
    /// Get all default keymaps
    pub fn defaults() -> Vec<DefaultKeymap> {
        vec![
            DefaultKeymap {
                action: "command_palette".to_string(),
                key: "p".to_string(),
                modifiers: "ctrl".to_string(),
                description: "Open command palette".to_string(),
            },
            DefaultKeymap {
                action: "new_connection".to_string(),
                key: "n".to_string(),
                modifiers: "ctrl+shift".to_string(),
                description: "Add new server".to_string(),
            },
            DefaultKeymap {
                action: "close_tab".to_string(),
                key: "w".to_string(),
                modifiers: "ctrl".to_string(),
                description: "Close current tab".to_string(),
            },
            DefaultKeymap {
                action: "next_tab".to_string(),
                key: "Tab".to_string(),
                modifiers: "ctrl".to_string(),
                description: "Switch to next tab".to_string(),
            },
            DefaultKeymap {
                action: "prev_tab".to_string(),
                key: "Tab".to_string(),
                modifiers: "ctrl+shift".to_string(),
                description: "Switch to previous tab".to_string(),
            },
            DefaultKeymap {
                action: "toggle_sidebar".to_string(),
                key: "b".to_string(),
                modifiers: "ctrl".to_string(),
                description: "Toggle sidebar visibility".to_string(),
            },
            DefaultKeymap {
                action: "search_servers".to_string(),
                key: "f".to_string(),
                modifiers: "ctrl".to_string(),
                description: "Search servers".to_string(),
            },
            DefaultKeymap {
                action: "open_sftp".to_string(),
                key: "e".to_string(),
                modifiers: "ctrl+shift".to_string(),
                description: "Open SFTP file manager".to_string(),
            },
            DefaultKeymap {
                action: "open_snippets".to_string(),
                key: "s".to_string(),
                modifiers: "ctrl+shift".to_string(),
                description: "Open snippets panel".to_string(),
            },
            DefaultKeymap {
                action: "disconnect".to_string(),
                key: "d".to_string(),
                modifiers: "ctrl+shift".to_string(),
                description: "Disconnect current session".to_string(),
            },
            DefaultKeymap {
                action: "copy_terminal".to_string(),
                key: "c".to_string(),
                modifiers: "ctrl+shift".to_string(),
                description: "Copy selected terminal text".to_string(),
            },
            DefaultKeymap {
                action: "paste_terminal".to_string(),
                key: "v".to_string(),
                modifiers: "ctrl+shift".to_string(),
                description: "Paste into terminal".to_string(),
            },
        ]
    }
}
